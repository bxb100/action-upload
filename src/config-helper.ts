import * as core from '@actions/core'

export class ConfigHelper {
  private readonly _provider: string
  private readonly _options: Record<string, string>
  private readonly _patterns: string[]

  constructor() {
    // provider
    this._provider = core.getInput('provider', {
      required: true,
      trimWhitespace: true
    })
    // provider options
    this._options = {}
    const provider_options = core
      .getMultilineInput('provider_options', {
        required: false
      })
      .filter((option: string) => option !== '')
    for (const option of provider_options) {
      const [key, value] = option.split('=')
      this._options[key] = value
    }
    // include file patterns
    this._patterns = core.getMultilineInput('include', {
      required: true,
      trimWhitespace: true
    })
  }

  get provider(): string {
    return this._provider
  }

  get options(): Record<string, string> {
    return this._options
  }

  get patterns(): string[] {
    return this._patterns
  }
}
