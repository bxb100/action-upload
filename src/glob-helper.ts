import * as core from '@actions/core'
import * as glob from '@actions/glob'
import * as os from 'os'

const patterSplit = (
  patterns: string[]
): {includes: string[]; excludes: string[]} => {
  const includes: string[] = []
  const excludes: string[] = []
  for (const pattern of patterns) {
    pattern.startsWith('!') ? excludes.push(pattern) : includes.push(pattern)
  }
  return {includes, excludes}
}

export declare interface PathSpec {
  dir: string
  path: string
  basename: string
  fsPath: string
}

export const includeFiles = async (patterns: string[]): Promise<PathSpec[]> => {
  const {includes, excludes} = patterSplit(patterns)
  // exclude must be set in the last
  const globber = await glob.create([...includes, ...excludes].join('\n'), {
    followSymbolicLinks: false,
    implicitDescendants: true,
    matchDirectories: false,
    omitBrokenSymbolicLinks: true
  })
  const searchPaths = globber.getSearchPaths()
  core.debug(`search paths: ${searchPaths}`)
  const paths: PathSpec[] = []
  let separate = '/'
  if (os.platform() === 'win32') {
    separate = '\\'
  }
  for await (const file of globber.globGenerator()) {
    // exclude .DS_Store
    if (file.endsWith('.DS_Store')) {
      continue
    }

    // according to the `@action/glob` rule, the getSearchPaths may return multiple
    for (const base of searchPaths) {
      core.debug(`base: ${base}; file: ${file}`)
      if (file.startsWith(base)) {
        let dir = file.substring(base.length, file.lastIndexOf(separate))
        if (dir) {
          // openDAL need the directory path end with '/'
          dir = `${dir}${separate}`
        }
        const basename = file.substring(file.lastIndexOf(separate) + 1)
        const path = file.substring(base.length + 1)
        paths.push({dir, path, basename, fsPath: file})
      }
    }
  }
  return paths
}
