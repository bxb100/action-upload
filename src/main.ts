import * as core from '@actions/core'
import * as fs from 'fs'
import {ConfigHelper} from './config-helper'
import {Operator} from 'opendal'
import {includeFiles} from './glob-helper'

async function run(): Promise<void> {
  try {
    const config = new ConfigHelper()
    core.debug(`provider options: ${JSON.stringify(config.options)}`)
    core.debug(`include patterns: ${JSON.stringify(config.patterns)}`)
    const op = new Operator(config.provider, config.options)
    core.startGroup(`Upload files to ${config.provider} start`)
    const pathSpec = await includeFiles(config.patterns)
    for (const spec of pathSpec) {
      core.debug(`upload file: ${spec.fsPath}`)
      core.info(`upload file: ${spec.path} tp ${spec.dir}`)
      // ensure the upload directory exists, relative path from search path
      if (spec.dir) {
        core.debug(`ensure the upload directory exists: ${spec.dir}`)
        await op.createDir(spec.dir)
      }
      // upload file
      await op.write(spec.path, fs.readFileSync(spec.fsPath))
    }
    core.endGroup()
  } catch (error) {
    if (error instanceof Error) core.setFailed(error.message)
  }
}

run()
