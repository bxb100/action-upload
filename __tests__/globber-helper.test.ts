import {test} from '@jest/globals'
import {includeFiles} from '../src/glob-helper'
import assert = require('assert')
import * as path from 'path'

test('test glob', async () => {
  const patterns = ['__tests__/']
  const file = await includeFiles(patterns)
  assert(file.length > 0)
  for (let pathSpec of file) {
    console.log(pathSpec)
    assert.equal(path.join(__dirname, pathSpec.path), pathSpec.fsPath)
  }
})

test('test double asterisk', async () => {
  const patterns = ['**/__tests__/']
  const file = await includeFiles(patterns)

  const patterns2 = ['**/__tests__/**']
  const file2 = await includeFiles(patterns2)

  assert.deepEqual(file, file2)

  for (let pathSpec of file) {
    assert(pathSpec.dir)
  }
}, 100000)
