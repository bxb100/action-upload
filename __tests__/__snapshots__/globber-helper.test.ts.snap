// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`github issue 110 - __test__ dir 1`] = `
[
  "__snapshots__/globber-helper.test.ts.snap",
  "config-helper.test.ts",
  "fixture/temp",
  "globber-helper.test.ts",
  "main.test.ts",
]
`;
